[FIXES JENKINS-47406] Milestone JOB_LOADED is now attained after cleaning up obsolete items

Vlatombe opened the pull request:
This allows `@Initializer(after=InitMilestone.JOB_LOADED)` implementations
to create items safely without risking them to be removed.

See [JENKINS-47406](https://issues.jenkins-ci.org/browse/JENKINS-47406).

I don't know how to build a unit test that would exhibit the problem. The idea would be to have an `@Initializer(after=JOB_LOADED)` instance forced to run first, creating an item, and checking that this item is still there once Jenkins is done with its initialization.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Fixed a race condition for Initializer implementations creating Items

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@oleg-nenashev I would love your review since I think you are the last one who touched the initialization sequence.

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

Vlatombe committed:
[FIXES JENKINS-47406] Milestone JOB_LOADED is now attained after cleaning up obsolete items

This allows @Initializer(after=InitMilestone.JOB_LOADED) implementations
to create items safely without risking them to be removed.

oleg-nenashev commented:
@Vlatombe it does not mean I want to touch it again :) Will take a look

oleg-nenashev commented:
@reviewbybees though

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev reviewed (approved):
🐝 from what I see. @reviewbybees done

oleg-nenashev commented:
@jenkinsci/core this change is ready to go, but I would suggest putting it on hold till the LTS baseline gets selected. Then we can backport it to .1 or .2 if it is stable enough

jglick commented:
> I don't know how to build a unit test that would exhibit the problem. The idea would be to have an `@Initializer(after=JOB_LOADED)` instance forced to run first, creating an item, and checking that this item is still there once Jenkins is done with its initialization.

All sounds testable to me. Create a test suite with such an initializer, no?

> I would suggest putting it on hold till the LTS baseline gets selected

-1. Train model: we merge changes when they are done. Ignore LTS baseline selection. If a given release looks bad in the field, we pick another. And we can always backport regression fixes easily enough.

oleg-nenashev commented:
According to @jglick's comment, needs a test case

oleg-nenashev commented:
@Vlatombe ping. Any plans to deliver a test?

Vlatombe commented:
@jglick I tried but as I can't control the execution order of initializers,
then I couldn't exhibit the problem in a reproducible way

Le sam. 4 nov. 2017 à 10:20, Oleg Nenashev <notifications@github.com> a
écrit :

> @Vlatombe <https://github.com/vlatombe> ping. Any plans to deliver a test?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/3078#issuecomment-341883139>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKdw4H0XLqOqgZQCeyKtYiZaqK4tWpxks5szCxmgaJpZM4P3FCF>
> .
>
-- 
Vincent

oleg-nenashev commented:
The code itself looks reasonable to me. I would propose merging it towards the next weekly. Next LTS baseline is in 3 months, so we will have enough time to handle issues if they happen. @jglick doe you agree?

jglick commented:
> I can't control the execution order of initializers

Well you can specify `before` and `after` but this is not enough to pin down the timing? What exactly is running concurrently that you cannot control? Perhaps `src/main/` code could be enhanced to make this aspect of startup more testable.

> I couldn't exhibit the problem in a reproducible way

Even a test which _sometimes_ fails without the patch would be valuable, I think.

I am hesitant to block a potentially valuable PR merely because it lacks a test, but on the other hand the correctness of the change is far from obvious from code inspection (to me at least), and this is pretty “core” functionality.

Vlatombe committed:
[JENKINS-47406] Add test with an initializer creating a job

Vlatombe commented:
@jglick I added a simple test checking that the basic behaviour this PR allows to do works.

I have been able to construct locally a failing test based on the previous impl, but it required to add additionnal named milestones in order to construct the right sequence of initializers that I was seeing in real world. This had much impact on the src/main code and I don't think this complexity is worth it just to verify that a potential race condition was happening. The new code leaves no room to this race condition anymore since the milestone `JOB_LOADED` is now attained once obsolete items have been removed.

jglick commented:
There seem to be a lot of test failures. Probably the new test broke other tests; we lack an equivalent to `@TestExtension` for `@Initializer`. Not sure if that is really solvable without a lot more work, since we also lack an equivalent to `GuiceExtensionAnnotation`.

BTW please do not rebase or force-push into a PR after it is filed. Just `git pull upstream master` or whatever as needed, and leave it to the merger of a PR to decide whether to squash or what. The build and comment history is all confused here.

Vlatombe commented:
Indeed, this looks like such a test is not fixable without additional support work.

jglick reviewed (dismissed):
Need to get tests passing again, even if that means just deleting or `@Ignore`ing the new test case, unfortunately.

Vlatombe committed:
Revert "[FIXES JENKINS-47406] Milestone JOB_LOADED is now attained after cleaning up obsolete items"

This reverts commit 79ce796b00fde11fda586c9c9e2b8e93a53e2b31.

jglick commented:
I think I know what you can do! Move the initializer to a plugin, compiled into a test resource `*.hpi`, then `@WithPlugin` to load it.

Vlatombe committed:
[JENKINS-47406] Reinstate the test with Jesse's solution

Vlatombe commented:
@jglick done.

oleg-nenashev reviewed (changes requested):
I am not sure I want to advertise `WithPlugin` in 2017, but IMHO it at least needs a link to the source code of the attached file in comments. Ideally the plugin need to be delivered using Maven (maybe overkill due to the permission)

jglick commented:
You seem to have reverted the actual fix here.

Vlatombe committed:
Reverted this commit by mistake

Revert "Revert "[FIXES JENKINS-47406] Milestone JOB_LOADED is now attained after cleaning up obsolete items""

This reverts commit 36c499c58c83427c7b8a6d3cc5902ec26732b2c7.

Vlatombe committed:
Add link to hpi sources

Vlatombe commented:
@jglick Indeed, seems like I had reverted the wrong commit
@oleg-nenashev do you have any alternative to `WithPlugin` for this use case?

I added the link to the hpi sources as requested.

oleg-nenashev commented:
@Vlatombe no alternative AFAIK. It could be possible to post plugins to Maven repo and download them, but it may require infra changes to avoid their appearance in update centers

oleg-nenashev commented:
Merge conflict after the recent security release

Vlatombe committed:
Merge branch 'master' into JENKINS-47406

Vlatombe commented:
@oleg-nenashev merge conflict fixed.

oleg-nenashev commented:
Re-requested review from @jglick

jglick commented:
> do you have any alternative to `WithPlugin` for this use case?

Not currently. @dwnusbaum & I had discussed a plugin builder test utility which would avoid the need to add binaries to the source tree for tests which just need to try installing a plugin. Would need special work to make such a system replace `@WithPlugin`, since that annotation runs before the test case starts.

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3070,10 +3071,10 @@ public void run(Reactor session) throws Exception {
         }
 
-        g.requires(JOB_LOADED).attains(COMPLETED).add("Cleaning up obsolete items deleted from the disk", new Executable() {
+        g.requires(loadJobs.toArray(new Handle[loadJobs.size()])).attains(COMPLETED).add("Cleaning up obsolete items deleted from the disk", new Executable() {
```
Seems hackish—really what we needed here is another milestone after `JOB_LOADED`, right? But I do not understand this issue very well.

Vlatombe replied:
I don't see why someone would want to depend on a milestone taking place in between job loading and obsolete job clean up - that is the current `JOB_LOADED` milestone. That's why I'm moving it after the obsolete job clean up has been done.  Code-wise, I agree it would be simpler to read with a new milestone, however I don't think there is any benefit to make such a milestone public.

jglick reviewed (commented):
Unclear to me whether JENKINS-47406 is really a legitimate use case. Seems like what you are really asking for is a new milestone.

Vlatombe committed:
Fix Mix-up after the previous merge

oleg-nenashev commented:
We will need to think about milestones later anyway. What hits me most is that we have no milestone for "CORE_LOGIC_COMPLETED" so that I cannot add initializers to run only between  `CORE_LOGIC_COMPLETED` and `COMPLETED`. Example: launching background tasks only after the core finishes its init steps (I would argue it's what's needed for Support Core as well: https://github.com/jenkinsci/support-core-plugin/blob/48faf43ac7d7e55d1f9115df66bc216dbaa0c232/src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java#L413). 

So I do see a reason for a new milestone to be public. But yes maybe it is a separate story (I will try to find a JIRA ticket, likely I have already created it). CC @Vlatombe

oleg-nenashev reviewed (approved):
🐝 a bit hackish, but ok

oleg-nenashev commented:
No other feedback, merging
